### PR TITLE
search: remove remote searching

### LINKED ISSUE #15107: Remove Github code search from `brew search`
### Verification

- [X] This issue's title and/or description do not reference a single formula e.g. `brew install wget`. If they do, open an issue at https://github.com/Homebrew/homebrew-core/issues/new/choose instead.

### Provide a detailed description of the proposed feature

Remove Github code search from our codebase and replace it with the Homebrew API. We only use this internally to search for formulas and casks with the `brew search` command and we already have that command set up to use the API right now for the core formula and cask repos.

The `Search.search_taps` and `Github.search_code` code should be removed. This will mean that we don't show casks that are not in the default cask repo in search results. This is less than ideal but can be handled in a few different ways. Either those repos can be merged together which is something we've already discussed in a few different places or we can add those to the API. The other option, of course, would be to just omit those from search results entirely unless you've tapped those repos.

### What is the motivation for the feature?

The idea here is that this doesn't work consistently unless you're using your own token and even in that case it's still inconsistent. This results in an extra network request for users that frequently doesn't return anything of value and also causes some of our online tests to be flaky. We should remove it.

### How will the feature be relevant to at least 90% of Homebrew users?

It will result in one less network request for most users and will help our CI be more consistent which will help everyone indirectly.

### What alternatives to the feature have been considered?

Only using code search if the user has a Github token to increase the likelihood that it works correctly.

### PR DESCRIPTION
**Pull request by Bo98**
Option A: Algolia
Option B (this PR): remove remote searching entirely

Given the use of our own JSON API, the need for remote searching has largely diminished as you can no longer "untap" Homebrew/cask.

However with both options we do lose the ability to search cask-versions/drivers/fonts when they are untapped, which was the reason the Algolia pull request was rejected so this PR is perhaps not enough of an improvement over the original proposal (though the current functionality is now broken without API tokens). If any ever get added to formulae.brew.sh (even if just as HTML with no JSON), then Algolia could cover them.

Option C is to use the GitHub files API which I could look at, though will require a number of API requests (per repository and perhaps pagination but I haven't checked).

Fixes https://github.com/Homebrew/brew/issues/15208
Closes #15107

### CODE REVIEW

File: Library/Homebrew/brew.sh [hunk 1]
```diff
@@ -810,6 +810,7 @@ fi
 AUTO_UPDATE_COMMANDS=(
   install
   outdated
+  search
```
Comment (apainintheneck) [inline Library/Homebrew/brew.sh:813]: Why is this necessary now?

**Review by apainintheneck — APPROVED** [review]
I think this make sense in the short-term because search seems to broken right now anyway. This should fix #15208.

Any of the other methods of adding the other cask taps to search can be handled in another PR. Either we can use https://github.com/Homebrew/brew/issues/15107 for that or we can create a new issue to remind ourselves to follow up on that.

File: Library/Homebrew/brew.sh (hunk 1 above)
Comment (Bo98) [reply Library/Homebrew/brew.sh:813]: Idea was this is the replacement for remote searching. The code we're removing always worked based on the latest remote state of the taps so doing an auto-update would keep similar behaviour.

Comment (MikeMcQuaid) [reply Library/Homebrew/brew.sh:813]: I'm conflicted on this. I see the benefit of this always being up-to-date. On the flip side, I've heard a bunch of people complain that "`brew search` is too slow" and I fear this may make it worse.

I'm tempted to say we merge without this and then be willing to add it in if we get any complaints or confusion over inconsistencies?

File: Library/Homebrew/search.rb [hunk 2]
```diff
@@ -42,51 +42,6 @@ def search_descriptions(string_or_regex, args, search_type: :desc)
       end
     end
 
-    def search_taps(query, silent: false)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/search.rb:45]: 🔥

**Review by MikeMcQuaid — COMMENTED** [review]
Great work here so far @Bo98. One tiny tweak and think this is good to go.

**Comment by MikeMcQuaid** [comment]
> Option B (this PR): remove remote searching entirely

I agree this is the best fit.

> However with both options we do lose the ability to search cask-versions/drivers/fonts when they are untapped, which was the reason the Algolia pull request was rejected so this PR is perhaps not enough of an improvement over the original proposal (though the current functionality is now broken without API tokens).

IMO this is yet another good reason to combine these taps. I still think we should be combining cask-fonts too, at least for those that are clearly used through analytics CC @Homebrew/cask.

**Comment by Bo98** [comment]
> I still think we should be combining cask-fonts too, at least for those that are clearly used through analytics CC @Homebrew/cask.

It's possible we could have formulae.brew.sh support for cask-fonts, I wouldn't say that's completely out of the question. Focus on the others first though.

**Comment by MikeMcQuaid** [comment]
> It's possible we could have formulae.brew.sh support for cask-fonts, I wouldn't say that's completely out of the question. Focus on the others first though.

I don't think it's worth it for a single tap. I think there's a host of issues with us having multiple official cask taps in the Homebrew organisation.

File: Library/Homebrew/brew.sh (hunk 1 above)
Comment (Bo98) [reply Library/Homebrew/brew.sh:813]: Speed shouldn't really be any slower than `brew outdated` after this change, unless the slowness is in DidYouMean or something.

So not really convinced about that as a reason, but I am happy to drop it anyway.

**Comment by MikeMcQuaid** [comment]
> We're talking about 2000 packages here though rather than a couple, which would increase the size of homebrew-cask by 50%.

I think using analytics can slim this down a bit but, post-sharding: I think this is still worthwhile.

> There's maintainability questions that would be best to discuss with the people who maintain it on a day-to-day basis.

I think maintainability is important but: so is user/contributor experience and the current level of discoverability makes both of those harder.

> formulae.brew.sh integration isn't really that tough of a task

Yeh, probably but but: we've got a host of other API-related issues affecting users that seem much higher priority to me: https://github.com/Homebrew/brew/issues?q=is%3Aissue+is%3Aopen+label%3A%22install+from+api%22

These are blocking our ability to say in https://github.com/Homebrew/brew/pull/14592 "hey, try enabling API support again".

**Comment by SMillerDev** [comment]
> There's maintainability questions that would be best to discuss with the people who maintain it on a day-to-day basis.

I only maintain the main cask repo because the others have increasingly hand-wavy rules so for me keeping them all in the same repo with the same rules would be a big improvement. Also note that https://github.com/Homebrew/homebrew-cask/issues/112102 has existed for 1,5 years at this point. Clearly nobody objects significantly.

**Comment by MikeMcQuaid** [comment]
> Let's keep that separate for now given we'd ideally look at those in the short term and cask-fonts in the medium term (sharding needs to happen before adding another 2000 casks, and versions & drivers should be tackled first).

My point was intended to be: I think all of those issues are much more important than fixing search for taps.

> Users don't really care as much about taps so maintainability is always going to be an important factor.

Users indirectly care about taps when they don't discover things because they don't know about non-default taps.

> I don't agree it's impossible to make a good user experience with separate taps

I don't think this either. I just think it requires a bunch of work that would be unnecessary for maintainers, contributors and users when you subdivide things into taps. This is why we stopped doing this (very) long ago for formulae.

File: Library/Homebrew/brew.sh (hunk 1 above)
Comment (Bo98) [reply Library/Homebrew/brew.sh]: Dropped now.

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
Thanks again @Bo98!
